### PR TITLE
test(otlp-common): expand test coverage for OTLP common exporter

### LINKED ISSUE #1667: Create separate gems for http, grpc export and a common gem they can rely on
The `exporter/otlp-http`, `exporter/otlp-grpc` and `exporter/otlp-common` directories in this repository reference tests to break out the existing `opentelemetry-exporter-otlp` gem into smaller chunks. Ruby doesn't currently have support for a gRPC exporter, and the behavior that would be shared between an HTTP and gRPC exporter would live in a common gem they both rely on.

Initial PRs related to this work: 
* https://github.com/open-telemetry/opentelemetry-ruby/pull/1179 
* https://github.com/open-telemetry/opentelemetry-ruby/pull/1180 

These exporters are not production ready as they currently exist. (See: https://github.com/open-telemetry/opentelemetry-ruby/issues/1337) 

Contributions welcome to add test coverage, error handling, and other work to make this production-ready.

### PR DESCRIPTION
**Pull request by arjun-rajappa**
Closes #1667
Expands test coverage for the OTLP common exporter and adds span flags support per OTLP specification.

Changes
Added build_span_flags method to set remote parent context flags on spans and links
Comprehensive test coverage for all private methods (as_otlp_span, as_otlp_status_code, as_otlp_span_kind, as_otlp_key_value, as_otlp_any_value)
Integration tests for complex spans with multiple resources and scopes
Enhanced error handling tests for encoding issues
Testing
All tests pass. Coverage expanded from basic functionality to comprehensive edge cases.

### CODE REVIEW

**Comment by kaylareopelle** [comment]
Hi @arjun-rajappa, thanks for your patience with my review. Could you take a look at the failing tests?

**Comment by thompson-tomo** [comment]
@kaylareopelle i have split out the new test cases to a seperate PR #2159 to enable this to be more focussed.

@arjun-rajappa it would be great if you could avoid mixing in format changes as it just ends up making it more time consuming to review.

**Comment by arjun-rajappa** [comment]
> Hi @arjun-rajappa, thanks for your patience with my review. Could you take a look at the failing tests?

I believe these failures are addressed already in https://github.com/open-telemetry/opentelemetry-ruby/pull/2074. As these are related to otlp-http. Once the other PR is merged in these should start passing.

**Comment by thompson-tomo** [comment]
I don't think that is the case as our main is passing so a change here is resulting in the failure.

Note common is a dependency of http hence why it is possible a change to common could lead to failures in http.

**Comment by arjun-rajappa** [comment]
These changes are closely related, as otlp-http relies on methods defined in otlp-common. As a result, changes to otlp-common can impact the tests in otlp-http. If needed, I can add the changes here or rebase this branch entirely on top of https://github.com/open-telemetry/opentelemetry-ruby/pull/2074.

**Comment by kaylareopelle** [comment]
> These changes are closely related, as otlp-http relies on methods defined in otlp-common. As a result, changes to otlp-common can impact the tests in otlp-http. If needed, I can add the changes here or rebase this branch entirely on top of https://github.com/open-telemetry/opentelemetry-ruby/pull/2074.

Any approach is fine here, as long as it's clear what was brought in and whether we need to adhere to a specific merge order. We can also wait to review this one until #2074 is merged too if that's easier.

**Comment by kaylareopelle** [comment]
Hi @arjun-rajappa, could you merge `main` into your branch so that we can see the tests run with the changes from #2074?

**Comment by kaylareopelle** [comment]
Hi @arjun-rajappa, it looks like there are some failures after merging `main` into this branch. The `translates all the things` tests are kind of tricky. Let me know if you want help fixing them.

**Comment by arjun-rajappa** [comment]
Hi @kaylareopelle ,
The previous failures have been fixed.

**Comment by thompson-tomo** [comment]
There is discussion in https://github.com/open-telemetry/opentelemetry-ruby/pull/2159 about the test cases, with that pr being purely the new tests added here.

@arjun-rajappa based on feedback of the tests, it would make things easier if this pr focussed on just  the new feature and the corresponding tests of that functionality with general tests added seperately

**Comment by kaylareopelle** [comment]
Hi @arjun-rajappa  - Could you take a look at the merge conflicts?

**Comment by thompson-tomo** [comment]
@arjun-rajappa note we decided in https://github.com/open-telemetry/opentelemetry-ruby/pull/2159 against testing the private methods and as such can u reflect that in this pr.

File: exporter/otlp-common/test/opentelemetry/exporter/otlp/common/common_test.rb [hunk 1]
```diff
@@ -148,52 +136,49 @@
...
+                        Opentelemetry::Proto::Trace::V1::SpanFlags::SPAN_FLAGS_CONTEXT_HAS_IS_REMOTE_MASK |
+                        1
                       )
                     ),
                     Opentelemetry::Proto::Trace::V1::Span.new(
-                      trace_id: trace_id,
-                      span_id: client_span_id,
-                      parent_span_id: child_span_id,
-                      name: 'client',
+                      trace_id: trace_id, span_id: client_span_id, parent_span_id: child_span_id, name: 'client',
```
Comment (kaylareopelle) [inline exporter/otlp-common/test/opentelemetry/exporter/otlp/common/common_test.rb:155]: Thoughts on reverting this spacing change?
```suggestion
                      trace_id: trace_id, 
                      span_id: client_span_id, 
                      parent_span_id: child_span_id, 
                      name: 'client',
```

File: exporter/otlp-common/test/opentelemetry/exporter/otlp/common/common_test.rb [hunk 2]
```diff
@@ -148,52 +136,49 @@
...
+                        Opentelemetry::Proto::Trace::V1::SpanFlags::SPAN_FLAGS_CONTEXT_HAS_IS_REMOTE_MASK |
+                        1
                       )
                     ),
                     Opentelemetry::Proto::Trace::V1::Span.new(
-                      trace_id: trace_id,
-                      span_id: consumer_span_id,
-                      parent_span_id: child_span_id,
-                      name: 'consumer',
+                      trace_id: trace_id, span_id: consumer_span_id, parent_span_id: child_span_id, name: 'consumer',
```
Comment (kaylareopelle) [inline exporter/otlp-common/test/opentelemetry/exporter/otlp/common/common_test.rb:168]:
```suggestion
                      trace_id: trace_id,
                      span_id: consumer_span_id,
                      parent_span_id: child_span_id,
                      name: 'consumer',
```

File: exporter/otlp-common/test/opentelemetry/exporter/otlp/common/common_test.rb [hunk 3]
```diff
@@ -229,30 +212,37 @@
...
-                    name: 'other_tracer'
-                  ),
+                  scope: Opentelemetry::Proto::Common::V1::InstrumentationScope.new(name: 'other_tracer'),
                   spans: [
                     Opentelemetry::Proto::Trace::V1::Span.new(
-                      trace_id: trace_id,
-                      span_id: server_span_id,
-                      parent_span_id: client_span_id,
-                      name: 'server',
+                      trace_id: trace_id, span_id: server_span_id, parent_span_id: client_span_id, name: 'server',
```
Comment (kaylareopelle) [inline exporter/otlp-common/test/opentelemetry/exporter/otlp/common/common_test.rb:236]:
```suggestion
                      trace_id: trace_id,
                      span_id: server_span_id,
                      parent_span_id: client_span_id,
                      name: 'server',
```

**Review by kaylareopelle — COMMENTED** [review]
Thanks for the updates, @arjun-rajappa. I added a few comments related to reverting a spacing change for readability. If I'm missing the reason why the spacing was changed, please let me know. 

In addition, given the current scope of this PR, what do you think about changing the title to something like: `feat(otlp-common): Add support for span flags` ? It won't trigger a release because we haven't released OTLP common yet, but it seems more accurate for the git history.

**Review by simi — APPROVED** [review]
Considering spacing issues and commit title will be fixed, looks ok to me.
